core/src/main/java/io/questdb/griffin/engine/functions/groupby/StdDevSampleDoubleGroupByFunction.java

```diff
@@ -0,0 +1,115 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.groupby;
+
+import io.questdb.cairo.ArrayColumnTypes;
+import io.questdb.cairo.ColumnType;
+import io.questdb.cairo.map.MapValue;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.engine.functions.DoubleFunction;
+import io.questdb.griffin.engine.functions.GroupByFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+import io.questdb.std.Numbers;
+import org.jetbrains.annotations.NotNull;
+
+public class StdDevSampleDoubleGroupByFunction extends DoubleFunction implements GroupByFunction, UnaryFunction {
+    private final Function arg;
+    private int valueIndex;
+
+    public StdDevSampleDoubleGroupByFunction(@NotNull Function arg) {
+        this.arg = arg;
+    }
+
+    @Override
+    public void computeFirst(MapValue mapValue, Record record) {
+        final double d = arg.getDouble(record);
+        if (Numbers.isFinite(d)) {
+            mapValue.putDouble(valueIndex, d);
+            mapValue.putDouble(valueIndex + 1, d*d);
+            mapValue.putLong(valueIndex + 2, 1L);
+        } else {
+            mapValue.putDouble(valueIndex, 0);
+            mapValue.putDouble(valueIndex + 1, 0);
+            mapValue.putLong(valueIndex + 2, 0);
+        }
+    }
+
+    @Override
+    public void computeNext(MapValue mapValue, Record record) {
+        final double d = arg.getDouble(record);
+        if (Numbers.isFinite(d)) {
+            mapValue.addDouble(valueIndex, d);
+            mapValue.addDouble(valueIndex + 1 , d*d);
+            mapValue.addLong(valueIndex + 2, 1L);
+        }
+    }
+
+    @Override
+    public void pushValueTypes(ArrayColumnTypes columnTypes) {
+        this.valueIndex = columnTypes.getColumnCount();
+        columnTypes.add(ColumnType.DOUBLE);
+        columnTypes.add(ColumnType.DOUBLE);
+        columnTypes.add(ColumnType.LONG);
+    }
+
+    @Override
+    public void setDouble(MapValue mapValue, double value) {
+        mapValue.putDouble(valueIndex, value);
+        mapValue.putDouble(valueIndex + 1, value*value);
+        mapValue.putLong(valueIndex + 2, 1L);
+    }
+
+    @Override
+    public void setNull(MapValue mapValue) {
+        mapValue.putDouble(valueIndex, Double.NaN);
+        mapValue.putDouble(valueIndex + 1, Double.NaN);
+        mapValue.putLong(valueIndex + 2, 0);
+    }
+
+    @Override
+    public Function getArg() {
+        return arg;
+    }
+
+    @Override
+    public double getDouble(Record rec) {
```

Comment:
nit: the following test fails due to precision loss. Would be nice to have some protection from FP precision loss and overflow problems.
```java
    @Test
    public void testStddevSampOverflow() throws Exception {
        assertMemoryLeak(() -> {
            compiler.compile("create table tbl1 as (select 100000000 x from long_sequence(1000000))", sqlExecutionContext);
            assertSql(
                    "select stddev_samp(x) from tbl1",
                    "stddev_samp\n0.0\n"
            );
        });
    }
```

Reply:
I put an overflow check, so that if the variance is negative to return 0.0. 
For precision loss of double arithmetic operations gets trickier. There are some workarounds but how do you think this should be handled?

For example this test:


```java
@Test
public void testStddevSampAllSameValues() throws Exception {
    assertMemoryLeak(() -> {
        compiler.compile("create table tbl1 as (select 17.2151921 x from long_sequence(100))", sqlExecutionContext);
        assertSql(
            "select stddev_samp(x) from tbl1",
            "stddev_samp\n0.0\n"
        );
    });
}
```

loses precision. 

How do you usually deal with this? Any suggestions or up to me?

Reply:
> How do you usually deal with this? Any suggestions or up to me?

Any suggestions are certainly welcome.

Iterative (recursive) calculation may be of some help, but I'm not sure if it would solve the problem:
https://www.heikohoffmann.de/htmlthesis/node134.html

We also had an issue with long overflow in the past (see #1824), but that's something different.

Reply:
Hi, thanks for the reply. I have checked about iterative calculation. Do you think something like this approach is suitable for this case? A more rough approach would be to consider zero anything that is very close to zero as is is a statistical calculation. Wouldn't sth like this be ok or we need a more concrete approach?

Reply:
I wouldn't care much about close to zero value rounding.

As for the recursive formula, the approach is certainly applicable to std dev:
https://math.stackexchange.com/questions/374881/recursive-formula-for-variance

Would be nice to know your opinion on whether it's worth it and what it would protect us from.

Reply:
> I wouldn't care much about close to zero value rounding.

Why not? Having only the same values multiple times which gives an stddev result that is not zero shouldn't be taken care of?
Btw I mentioned the near zero issue just for the specific case of the test above. 

For example postgres behaves quite similar.

```sql
create table tbl1(x double precision);

insert into tbl1 select 17.2151921 from generate_series(1,100);
```


```sql
select stddev_samp(x) from tbl3;

      stddev_samp       
------------------------
 1.0634742720642203e-14
```


As of generally for computing standard deviation:

So stddev calculation involves some arithmetic computation issues and there are trade-offs about using an online calculation (incrementally as data arrive) or an offline calculation when you have all data (also mean value) beforehand. Since we need an online calculation and not the classic formula when mean value is known, a recursive approach would help us avoid miscalculations.

Using the sum-of-squares appears to be a red flag for some calculations.
If the x‘s are large and the differences between them small, direct evaluation of the equation above would require computing a small number as the difference of two large numbers, a red flag for numerical computing

So it seems that the sum-of-squares loses precision that can end up even in a negative variance result due to precision loss. Having an approach like Welford’s method is shown to be more accurate ([link](https://www.johndcook.com/blog/2008/09/28/theoretical-explanation-for-numerical-results/)).

The last git push has the implementation of Welford’s method and the tests pass with high accuracy (better than postgres 🙂 as shown in the postgres example above )